File: 03_patterns/swift/mvvm_clean_arc/MVVM/Services/TaskService.swift
import Foundation

final class TaskService {
    private let store: SharedTaskStore

    init(store: SharedTaskStore = .shared) {
        self.store = store
    }

    func fetchTasks() -> [TaskItem] {
        store.fetchTasks().map(\.asTaskItem)
    }

    func createTask(title: String) -> [TaskItem] {
        store.createTask(title: title).map(\.asTaskItem)
    }

    func deleteTask(id: UUID) -> [TaskItem] {
        store.deleteTask(id: id).map(\.asTaskItem)
    }
}

private extension SharedTaskRecord {
    var asTaskItem: TaskItem {
        TaskItem(id: id, title: title)
    }
}

File: 03_patterns/flutter/lib/mvvm/services/task_service.dart
import '../../shared/models/task_item.dart';
import '../../shared/storage/shared_task_store.dart';

class TaskService {
  TaskService({SharedTaskStore? store}) : _store = store ?? SharedTaskStore.shared;

  final SharedTaskStore _store;

  List<TaskItem> fetchTasks() {
    return _store.fetchTasks().map(_asTaskItem).toList();
  }

  List<TaskItem> createTask({required String title}) {
    return _store.createTask(title: title).map(_asTaskItem).toList();
  }

  List<TaskItem> deleteTask({required String id}) {
    return _store.deleteTask(id: id).map(_asTaskItem).toList();
  }

  TaskItem _asTaskItem(SharedTaskRecord record) {
    return TaskItem(id: record.id, title: record.title);
  }
}
